### PR TITLE
Allow shareable resources for admission control plugins.

### PR DESCRIPTION
**Pull request by hodovska**
Changes allow admission control plugins to share resources. This is done via new PluginInitialization structure. The structure can be extended for other resources, for now it is an shared informer for namespace plugins (NamespiceLifecycle, NamespaceAutoProvisioning, NamespaceExists).

If a plugins needs some kind of shared resource e.g. client, the client shall be added to PluginInitializer and Wants methods implemented to every plugin which will use it.

### CODE REVIEW

**Comment by ncdc** [comment]
@hodovska could you please fill in the description with some information about what this does?

**Comment by ncdc** [comment]
FYI @smarterclayton

File: pkg/admission/types.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+*/
+
+package admission
+
+import "k8s.io/kubernetes/pkg/controller/framework"
+
+// Validator should be implemented by admission plugins that can validate themselves
+// after initialization has happened.
+type Validator interface {
+	Validate() error
```
Comment (ncdc) [inline pkg/admission/types.go]: Godoc on the function too, please

File: pkg/admission/types.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+// WantsNamespaceInformer should be implemented by admission plugins that need
+// a NamespaceInformer
+type WantsNamespaceInformer interface {
+	SetNamespaceInformer(framework.SharedIndexInformer)
```
Comment (ncdc) [inline pkg/admission/types.go]: Godoc please

File: plugin/pkg/admission/namespace/autoprovision/admission.go [hunk 3]
```diff
@@ -75,28 +76,19 @@ func (p *provision) Admit(a admission.Attributes) (err error) {
...
 		client:  c,
-		store:   store,
 	}
 }
+
+func (p *provision) SetNamespaceInformer(c framework.SharedIndexInformer) {
+	p.informer = c
+}
+
+func (p *provision) Validate() error {
```
Comment (ncdc) [inline plugin/pkg/admission/namespace/autoprovision/admission.go]: Could you please add a test case for this function?

File: pkg/admission/chain.go [hunk 4]
```diff
@@ -23,15 +23,20 @@ type chainAdmissionHandler []Interface
 
 // NewFromPlugins returns an admission.Interface that will enforce admission control decisions of all
 // the given plugins.
-func NewFromPlugins(client clientset.Interface, pluginNames []string, configFilePath string) Interface {
+func NewFromPlugins(client clientset.Interface, pluginNames []string, configFilePath string, plugInit PluginInitializer) (Interface, error) {
```
Comment (smarterclayton) [inline pkg/admission/chain.go]: Can you split this change specifically into its own commit - this appears to be the only API change here, and it's better it be separate.

**Comment by hodovska** [comment]
@nikhiljindal change in the federation-apiserver docs is now preventing tests from succeeding, could you please specify relevance of the changes to this PR?

**Comment by nikhiljindal** [comment]
@hodovska Sorry I dont understand, all tests seem to be passing here.

**Comment by nikhiljindal** [comment]
Unassigning myself since I dont think I will be able to get to it till 1.3 is out.
I am not the best reviewer for this anyway.

@smarterclayton you can take this up?
cc @wojtek-t

**Comment by smarterclayton** [comment]
Yeah I can cover it.

File: cmd/kube-apiserver/app/server.go [hunk 5]
```diff
@@ -243,7 +245,15 @@ func Run(s *options.APIServer) error {
 	if err != nil {
 		glog.Errorf("Failed to create clientset: %v", err)
 	}
-	admissionController := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile)
+
+	namespaceInformer := informers.CreateSharedNamespaceIndexInformer(client, 5*time.Minute)
```
Comment (derekwaynecarr) [inline cmd/kube-apiserver/app/server.go]: I think this ResyncPeriod should be fuzzed similar to controller MinResyncPeriod

File: federation/cmd/federation-apiserver/app/server.go [hunk 6]
```diff
@@ -119,8 +121,14 @@ func Run(s *genericoptions.ServerRunOptions) error {
 	if err != nil {
 		glog.Errorf("Failed to create clientset: %v", err)
 	}
-	admissionController := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile)
+	namespaceInformer := informers.CreateSharedNamespaceIndexInformer(client, 5*time.Minute)
```
Comment (derekwaynecarr) [inline federation/cmd/federation-apiserver/app/server.go]: same here, it should be a fuzzed similar to controller minResyncPeriod

File: pkg/admission/init.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"reflect"
+)
+
+// PluginInitializer is used for Initialization of shareable resources between admission plugins
+// After Initialization the resources have to be set separately
+type PluginInitializer interface {
+	Initialize(plugins []Interface)
+	SetNamespaceInformer(namespaceInformer framework.SharedIndexInformer)
```
Comment (derekwaynecarr) [inline pkg/admission/init.go]: can we change this to use SetInformer(type, informer) ?

File: pkg/admission/init.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"reflect"
+)
+
+// PluginInitializer is used for Initialization of shareable resources between admission plugins
+// After Initialization the resources have to be set separately
+type PluginInitializer interface {
+	Initialize(plugins []Interface)
```
Comment (derekwaynecarr) [inline pkg/admission/init.go]: nit: godoc

File: pkg/admission/init.go (hunk 7 above)
Comment (derekwaynecarr) [inline pkg/admission/init.go]: nit: godoc

File: pkg/admission/init.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+func (i *pluginInitializer) SetNamespaceInformer(namespaceInformer framework.SharedIndexInformer) {
+	i.informers[reflect.TypeOf(&api.Namespace{})] = namespaceInformer
+}
+
+// Initialize will check the initialization interfaces implemented by each plugin
+// and provide the appropriate initialization data
+func (i *pluginInitializer) Initialize(plugins []Interface) {
+	for _, plugin := range plugins {
+		if wantsNamespaceInformer, ok := plugin.(WantsNamespaceInformer); ok {
+			wantsNamespaceInformer.SetNamespaceInformer(i.informers[reflect.TypeOf(&api.Namespace{})])
```
Comment (derekwaynecarr) [inline pkg/admission/init.go]: doesnt need to change now, but it would be good if we can somehow associate a wants to a setter and inject it into this object so folks don't need to change this code to support other wants...

File: pkg/admission/types.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+import "k8s.io/kubernetes/pkg/controller/framework"
+
+// Validator holds Validate functions, which are responsible for validation of initialized shared resources
+// and should be implemented on admission plugins
+type Validator interface {
+	Validate() error
+}
+
+// WantsNamespaceInformer defines a function witch sets NamespaceInformer for admission plugins that need it
```
Comment (derekwaynecarr) [inline pkg/admission/types.go]: s/witch/which

**Comment by deads2k** [comment]
In openshift I think we've settled on an interface like this to pass around: https://github.com/openshift/origin/blob/master/pkg/controller/shared/shared_informer.go#L16 .  It allows for lazy registration, listwatch customization to handle direct-to-etcd list/watching, core vs. "normal" informers, and easy access to listers for components that only  need a cache.

Could you incorporate something similar?

**Comment by hodovska** [comment]
@deads2k @ncdc please review

File: pkg/controller/framework/informers/factory.go [hunk 11]
```diff
@@ -27,6 +28,56 @@ import (
...
+func (s *sharedInformerFactory) Start(resyncPeriod time.Duration) {
+	s.informers[reflect.TypeOf(&api.Pod{})] = CreateSharedPodIndexInformer(s.client, resyncPeriod)
+	s.informers[reflect.TypeOf(&api.Node{})] = CreateSharedNodeIndexInformer(s.client, resyncPeriod)
+	s.informers[reflect.TypeOf(&api.Namespace{})] = CreateSharedNamespaceIndexInformer(s.client, resyncPeriod)
+	s.informers[reflect.TypeOf(&api.PersistentVolumeClaim{})] = CreateSharedPVCIndexInformer(s.client, resyncPeriod)
+	s.informers[reflect.TypeOf(&api.PersistentVolume{})] = CreateSharedPVIndexInformer(s.client, resyncPeriod)
+}
+
+func (f *sharedInformerFactory) Pod() framework.SharedIndexInformer {
+	return f.informers[reflect.TypeOf(&api.Pod{})]
```
Comment (deads2k) [inline pkg/controller/framework/informers/factory.go]: You should be calling this method before you've called `.Start` (so you can register ahead of time), which means this would most often return `nil`.

If you look at the downsteam implementation, you'll notice that we're lazily initializing these so that we only create watches for things that are requested and it also ensures that we never get nil.

File: pkg/controller/framework/informers/factory.go [hunk 12]
```diff
@@ -27,6 +28,56 @@ import (
 	"k8s.io/kubernetes/pkg/watch"
 )
 
+type SharedInformerFactory interface {
+	Start(time.Duration)
+
+	Pod() framework.SharedIndexInformer
```
Comment (deads2k) [inline pkg/controller/framework/informers/factory.go]: The downstream implementation has this return an interface that gives you a `Lister` or an `Informer`.  The distinction is valuable since many locations are only interested in having the cache and we can start shaping the `Lister` API to behave more like a `ClientSet` API so the code reads consistently.

File: pkg/controller/framework/informers/factory.go [hunk 13]
```diff
@@ -27,6 +28,56 @@ import (
...
+}
+
+func NewSharedInformerFactory(client clientset.Interface) SharedInformerFactory {
+	return &sharedInformerFactory{
+		client:    client,
+		informers: make(map[reflect.Type]framework.SharedIndexInformer),
+	}
+}
+
+func (s *sharedInformerFactory) Start(resyncPeriod time.Duration) {
```
Comment (deads2k) [inline pkg/controller/framework/informers/factory.go]: This should be re-entrant to handle cases where it gets called more than once.  That can happen in bootstrapping cases and in cases where we have late joiners.

File: pkg/controller/framework/informers/factory.go [hunk 14]
```diff
@@ -25,8 +26,60 @@ import (
...
+func NewSharedInformerFactory(client clientset.Interface, defaultResync time.Duration) SharedInformerFactory {
+	return &sharedInformerFactory{
+		client:        client,
+		defaultResync: defaultResync,
+		informers:     make(map[reflect.Type]framework.SharedIndexInformer),
+	}
+}
+
+func (s *sharedInformerFactory) Start(stopCh <-chan struct{}) {
+	for _, informer := range s.informers {
```
Comment (deads2k) [inline pkg/controller/framework/informers/factory.go]: acquire the lock

File: pkg/controller/framework/informers/factory.go [hunk 15]
```diff
@@ -25,8 +26,60 @@ import (
 	"k8s.io/kubernetes/pkg/controller/framework"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/watch"
+	"sync"
 )
 
+type SharedInformerFactory interface {
+	Start(stopCh <-chan struct{})
+
+	Pod() PodInformer
```
Comment (deads2k) [inline pkg/controller/framework/informers/factory.go]: plural to look closer to the client API.

File: pkg/controller/framework/informers/informers.go [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+	"reflect"
+)
+
+type PodInformer interface {
+	Informer() framework.SharedIndexInformer
+	Indexer() cache.Indexer
```
Comment (deads2k) [inline pkg/controller/framework/informers/informers.go]: I can see dropping this.  I think we want to push people to the `Lister` if they aren't adding indexes and towards the `Informer` if they are.

File: pkg/controller/framework/informers/informers.go [hunk 17]
```diff
@@ -0,0 +1,277 @@
+/*
```
Comment (deads2k) [inline pkg/controller/framework/informers/informers.go]: This file probably ought to be `core` to match the client library.

File: pkg/controller/framework/informers/informers.go [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+func (f *namespaceInformer) Informer() framework.SharedIndexInformer {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	informerObj := &api.Namespace{}
+	informerType := reflect.TypeOf(informerObj)
+	informer, exists := f.informers[informerType]
+	if exists {
+		return informer
+	}
+	informer = framework.NewSharedIndexInformer(
```
Comment (deads2k) [inline pkg/controller/framework/informers/informers.go]: This seems to make `CreateSharedNamespaceIndexInformer` superfluous.  I think this construction here is what we want.  It makes it very easy to share the informers and will prevent accidents.

File: pkg/admission/types.go [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+import "k8s.io/kubernetes/pkg/controller/framework"
+
+// Validator holds Validate functions, which are responsible for validation of initialized shared resources
+// and should be implemented on admission plugins
+type Validator interface {
+	Validate() error
+}
+
+// WantsNamespaceInformer defines a function which sets NamespaceInformer for admission plugins that need it
+type WantsNamespaceInformer interface {
```
Comment (deads2k) [inline pkg/admission/types.go]: One problem we have downstream is remembering to force people to have a `Validate` method.  How about having `WantsNamespaceInformer` overlap by including `Validator`?

File: cmd/kube-apiserver/app/server.go [hunk 20]
```diff
@@ -243,7 +245,13 @@ func Run(s *options.APIServer) error {
 	if err != nil {
 		glog.Errorf("Failed to create clientset: %v", err)
 	}
-	admissionController := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile)
+	shared_informers := informers.NewSharedInformerFactory(client, 10*time.Minute)
```
Comment (ncdc) [inline cmd/kube-apiserver/app/server.go]: No underscores - call this sharedInformers instead

File: cmd/kube-apiserver/app/server.go [hunk 21]
```diff
@@ -243,7 +245,13 @@ func Run(s *options.APIServer) error {
...
+	pluginInit := admission.NewPluginInitializer(shared_informers)
```
Comment (ncdc) [inline cmd/kube-apiserver/app/server.go]: I think avoiding abbreviation is better - pluginInitializer

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 22]
```diff
@@ -196,22 +194,14 @@ func Run(s *options.CMServer) error {
...
-	pvcInformer := informers.CreateSharedPVCIndexInformer(clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "pvc-informer")), ResyncPeriod(s)())
-	pvInformer := informers.CreateSharedPVIndexInformer(clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "pv-informer")), ResyncPeriod(s)())
-	informers := map[reflect.Type]framework.SharedIndexInformer{}
-	informers[reflect.TypeOf(&api.Pod{})] = podInformer
-	informers[reflect.TypeOf(&api.Node{})] = nodeInformer
-	informers[reflect.TypeOf(&api.PersistentVolumeClaim{})] = pvcInformer
-	informers[reflect.TypeOf(&api.PersistentVolume{})] = pvInformer
-
-	go endpointcontroller.NewEndpointController(podInformer, clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "endpoint-controller"))).
+	sharedinformers := informers.NewSharedInformerFactory(clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "sharedinformers")), ResyncPeriod(s)())
```
Comment (ncdc) [inline cmd/kube-controller-manager/app/controllermanager.go]: capital i - sharedInformers

Comment (ncdc) [inline cmd/kube-controller-manager/app/controllermanager.go]: And the user agent should probably be shared-informers

File: federation/cmd/federation-apiserver/app/server.go [hunk 23]
```diff
@@ -119,8 +121,13 @@ func Run(s *genericoptions.ServerRunOptions) error {
 	if err != nil {
 		glog.Errorf("Failed to create clientset: %v", err)
 	}
-	admissionController := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile)
+	shared_informers := informers.NewSharedInformerFactory(client, 10*time.Minute)
```
Comment (ncdc) [inline federation/cmd/federation-apiserver/app/server.go]: sharedInformers

File: federation/cmd/federation-apiserver/app/server.go [hunk 24]
```diff
@@ -119,8 +121,13 @@ func Run(s *genericoptions.ServerRunOptions) error {
...
+	pluginInit := admission.NewPluginInitializer(shared_informers)
```
Comment (ncdc) [inline federation/cmd/federation-apiserver/app/server.go]: pluginInitializer

File: federation/cmd/federation-apiserver/app/server.go [hunk 25]
```diff
@@ -119,8 +121,13 @@ func Run(s *genericoptions.ServerRunOptions) error {
...
 
+	admissionController, err := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile, pluginInit)
+	if err != nil {
+		glog.Errorf("Failed to initialize plugins: %v", err)
```
Comment (ncdc) [inline federation/cmd/federation-apiserver/app/server.go]: This should probably be fatal, right?

File: pkg/admission/init.go [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"k8s.io/kubernetes/pkg/controller/framework/informers"
+)
+
+// PluginInitializer is used for Initialization of shareable resources between admission plugins
```
Comment (ncdc) [inline pkg/admission/init.go]: Add period at end of line

File: pkg/admission/init.go [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+// After Initialization the resources have to be set separately
```
Comment (ncdc) [inline pkg/admission/init.go]: lowercase initialization, period at end of line

File: pkg/admission/init.go [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+type PluginInitializer interface {
+	Initialize(plugins []Interface)
+}
+
+type pluginInitializer struct {
+	informers informers.SharedInformerFactory
+}
+
+// NewPluginInitializer constructs new instance of PluginInitializer
+func NewPluginInitializer(shared_informers informers.SharedInformerFactory) PluginInitializer {
```
Comment (ncdc) [inline pkg/admission/init.go]: sharedInformers

**Comment by ncdc** [comment]
FYI the license header format changed after you started this work. You'll need to remove "All rights reserved" - see https://github.com/kubernetes/kubernetes/blob/master/hack/boilerplate/boilerplate.go.txt for the latest template

**Comment by ncdc** [comment]
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/pkg/admission/init.go

File: plugin/pkg/admission/namespace/autoprovision/admission.go [hunk 29]
```diff
@@ -59,7 +60,7 @@ func (p *provision) Admit(a admission.Attributes) (err error) {
 		},
 		Status: api.NamespaceStatus{},
 	}
-	_, exists, err := p.store.Get(namespace)
+	_, exists, err := p.informer.GetStore().Get(namespace)
```
Comment (deads2k) [inline plugin/pkg/admission/namespace/autoprovision/admission.go]: doing this suggests that their lister is incomplete, let's get a followup to fix it.

File: pkg/controller/framework/informers/factory.go [hunk 30]
```diff
@@ -25,8 +26,73 @@ import (
...
+	}
+}
+
+// Start initializes all requested informers.
+func (s *sharedInformerFactory) Start(stopCh <-chan struct{}) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, informer := range s.informers {
+		go informer.Run(stopCh)
```
Comment (deads2k) [inline pkg/controller/framework/informers/factory.go]: We found upstream that we needed to call `.Start` multiple times in some cases.  We can add the power in a followup.

File: plugin/pkg/admission/namespace/exists/admission.go [hunk 31]
```diff
@@ -60,7 +60,7 @@ func (e *exists) Admit(a admission.Attributes) (err error) {
 		},
 		Status: api.NamespaceStatus{},
 	}
-	_, exists, err := e.store.Get(namespace)
+	_, exists, err := e.informer.GetStore().Get(namespace)
```
Comment (deads2k) [inline plugin/pkg/admission/namespace/exists/admission.go]: another incomplete lister, I'd guess.

File: plugin/pkg/admission/namespace/lifecycle/admission.go [hunk 32]
```diff
@@ -65,7 +64,7 @@ func (l *lifecycle) Admit(a admission.Attributes) (err error) {
 		// this will cause a live lookup of the namespace to get its latest state even
 		// before the watch notification is received.
 		if a.GetOperation() == admission.Delete {
-			l.store.Delete(&api.Namespace{
+			l.informer.GetStore().Delete(&api.Namespace{
```
Comment (deads2k) [inline plugin/pkg/admission/namespace/lifecycle/admission.go]: No.  This can't be done.  This will mutate the cache for everyone.  This can't share an informer yet.  You'll need a followup issue to create an overlay on the cache.

File: pkg/admission/types.go [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
```
Comment (deads2k) [inline pkg/admission/types.go]: You probably want this to be `WantsInformerFactory`.

File: plugin/pkg/admission/namespace/lifecycle/admission.go [hunk 34]
```diff
@@ -65,7 +64,7 @@ func (l *lifecycle) Admit(a admission.Attributes) (err error) {
 		// this will cause a live lookup of the namespace to get its latest state even
 		// before the watch notification is received.
 		if a.GetOperation() == admission.Delete {
-			l.store.Delete(&api.Namespace{
+			l.informerFactory.Namespaces().Informer().GetStore().Delete(&api.Namespace{
```
Comment (deads2k) [inline plugin/pkg/admission/namespace/lifecycle/admission.go]: This still isn't safe.  How about reverting the change for this admission plugin and circle back to it separately.

**Comment by deads2k** [comment]
Looks like that was the last of my comments.  Let's get a squash.

File: cmd/kube-apiserver/app/server.go [hunk 35]
```diff
@@ -58,6 +57,7 @@ import (
 	"k8s.io/kubernetes/pkg/registry/rolebinding"
 	rolebindingetcd "k8s.io/kubernetes/pkg/registry/rolebinding/etcd"
 	"k8s.io/kubernetes/pkg/serviceaccount"
+	"time"
```
Comment (ncdc) [inline cmd/kube-apiserver/app/server.go]: I wouldn't have expected this import to move

File: federation/cmd/federation-apiserver/app/server.go [hunk 36]
```diff
@@ -119,8 +121,13 @@ func Run(s *genericoptions.ServerRunOptions) error {
 	if err != nil {
 		glog.Errorf("Failed to create clientset: %v", err)
 	}
-	admissionController := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile)
+	sharedInformers := informers.NewSharedInformerFactory(client, 10*time.Minute)
```
Comment (deads2k) [inline federation/cmd/federation-apiserver/app/server.go]: spawn an issue to make this somethign on struct that we can pass in.

**Comment by deads2k** [comment]
I must be missing where you're calling `InformerFactory.Start`.  Are you sure the admission plugins are being tested?

If @ncdc sees it, this looks like a good starting point to me.  I'd like to see followup issues.

File: federation/cmd/federation-apiserver/app/server.go (hunk 36 above)
Comment (ncdc) [inline federation/cmd/federation-apiserver/app/server.go]: You mean on ServerRunOptions?

**Comment by ncdc** [comment]
@deads2k you're right - it looks like it's only called in the unit tests. We just need to do it after the calls to `informers.NewSharedInformerFactory`, right? @hodovska could you please make sure to do this?

**Comment by deads2k** [comment]
You should start the informers just befor you launch the server

File: federation/cmd/federation-apiserver/app/server.go (hunk 36 above)
Comment (deads2k) [inline federation/cmd/federation-apiserver/app/server.go]: > You mean on ServerRunOptions?

Hrm, now I remember why we're unable to use these paths to build the server downstream.  We have come in a layer deeper to provide a completely separate admission chain.

**Comment by ncdc** [comment]
@k8s-bot e2e test this issue: #29308

**Comment by ncdc** [comment]
@deads2k Start() calls added

File: plugin/pkg/admission/namespace/autoprovision/admission.go [hunk 37]
```diff
@@ -59,7 +60,7 @@ func (p *provision) Admit(a admission.Attributes) (err error) {
 		},
 		Status: api.NamespaceStatus{},
 	}
-	_, exists, err := p.store.Get(namespace)
+	_, exists, err := p.informerFactory.Namespaces().Informer().GetStore().Get(namespace)
```
Comment (deads2k) [inline plugin/pkg/admission/namespace/autoprovision/admission.go]: Followup issue to wait for the cache to sync.  This is a pre-existing bug.

File: plugin/pkg/admission/namespace/exists/admission.go [hunk 38]
```diff
@@ -41,10 +39,12 @@ func init() {
 // It is useful in deployments that want to enforce pre-declaration of a Namespace resource.
 type exists struct {
 	*admission.Handler
-	client clientset.Interface
-	store  cache.Store
+	client          clientset.Interface
+	informerFactory informers.SharedInformerFactory
```
Comment (deads2k) [inline plugin/pkg/admission/namespace/exists/admission.go]: When you sweep for `HasSynced` go ahead update these to hold only the information they need.  You still init based on the factory, but you store the values you need.

**Comment by deads2k** [comment]
It's a good starting point (functionally equivalent), lets merge it and follow up with issues and fixes for:

 - [ ] refactor the `CreateSharedXXXIndexInformer` and lazy informerfactory gets so that one calls the other.
 - [ ] allow `.Start` to be called multiple times
 - [ ] wait for caches to sync in admission plugins before admitting (check the openshift clusterquota admission for an example)
 - [ ] update the admission plugins to hold the values they need from the informerfactory when initialized
 - [ ] update the controllers to use this interface.

**Comment by ncdc** [comment]
Not sure why the e2e is listed as failed because the gs bucket only shows started.json and no artifacts. @ixdy @spxtr @pwittrock ?

@k8s-bot e2e test this issue: #IGNORE

**Comment by ncdc** [comment]
GKE smoke e2e failure is #27462 (again!)

**Comment by ncdc** [comment]
@k8s-bot e2e test this issue: #27462

**Comment by ncdc** [comment]
@k8s-bot e2e test this issue: #29308
